WordPress/src/main/java/org/wordpress/android/ui/uploads/LocalDraftUploadStarter.kt

```diff
@@ -1,41 +1,138 @@
+import org.wordpress.android.viewmodel.helpers.ConnectionStatus
 import javax.inject.Inject
 import javax.inject.Named
+import javax.inject.Singleton
+import kotlin.coroutines.CoroutineContext
 
 /**
- * Provides a way to find and upload all local drafts.
+ * Automatically uploads local drafts.
+ *
+ * Auto-uploads happen when the app is placed in the foreground or when the internet connection is restored. In
+ * addition to this, call sites can also request an immediate execution by calling [upload].
+ *
+ * The method [activateAutoUploading] must be called once, preferably during app creation, for the auto-uploads to work.
  */
+@Singleton
 class LocalDraftUploadStarter @Inject constructor(
     /**
      * The Application context
      */
     private val context: Context,
     private val postStore: PostStore,
+    private val siteStore: SiteStore,
+    @Named(BG_THREAD) private val bgDispatcher: CoroutineDispatcher,
+    @Named(IO_THREAD) private val ioDispatcher: CoroutineDispatcher,
+    private val uploadServiceFacade: UploadServiceFacade,
+    private val networkUtilsWrapper: NetworkUtilsWrapper,
+    private val connectionStatus: LiveData<ConnectionStatus>
+) : CoroutineScope {
+    private val job = Job()
+
+    override val coroutineContext: CoroutineContext get() = job + bgDispatcher
+
     /**
-     * The Coroutine dispatcher used for querying in FluxC.
+     * The hook for making this class automatically launch uploads whenever the app is placed in the foreground.
      */
-    @Named(BG_THREAD) private val bgDispatcher: CoroutineDispatcher,
-    private val networkUtilsWrapper: NetworkUtilsWrapper
-) {
-    fun uploadLocalDrafts(scope: CoroutineScope, site: SiteModel) = scope.launch(bgDispatcher) {
+    private val processLifecycleObserver = object : LifecycleObserver {
+        @OnLifecycleEvent(Event.ON_START)
+        fun onAppComesFromBackground() {
+            queueUploadFromAllSites()
+        }
+    }
+
+    /**
+     * Activates the necessary observers for this class to start auto-uploading.
+     *
+     * This must be called during [org.wordpress.android.WordPress]' creation like so:
+     *
+     * ```
+     * mLocalDraftUploadStarter.activateAutoUploading(ProcessLifecycleOwner.get())
+     * ```
+     */
+    fun activateAutoUploading(processLifecycleOwner: ProcessLifecycleOwner) {
+        // Since this class is meant to be a Singleton, it should be fine (I think) to use observeForever in here.
+        // We're skipping the first emitted value because the processLifecycleObserver below will also trigger an
+        // immediate upload.
+        connectionStatus.skip(1).observeForever {
+            queueUploadFromAllSites()
+        }
+
+        processLifecycleOwner.lifecycle.addObserver(processLifecycleObserver)
+    }
+
+    private fun queueUploadFromAllSites() = launch {
+        val sites = siteStore.sites
+        try {
+            checkConnectionAndUpload(sites = sites)
+        } catch (e: Exception) {
```

Comment:
How would you feel about catching a more specific exception type. I believe we should follow the Fail Fast principle and avoid hiding unexpected exceptions. Wdyt?

Reply:
I initially thought of failing fast for this class. But realized that it's a class that will be executed on multiple scenarios like app start, app foreground, etc. If the crash is consistent, users would not be able to use the app anymore. 

But if you feel that crashes are unlikely, we can remove the exception handling. I think they're unlikely but I opted to be on the safe side.

Reply:
As for the specific exception type handling, I only know of the documented `@throws` in `context.startService`. What exceptions should we expect from FluxC queries? In only saw `RuntimeException` from WellSql which is not as specific as I wanted it to be. 😬

Reply:
>I initially thought of failing fast for this class. But realized that it's a class that will be executed on multiple scenarios like app start, app foreground, etc. If the crash is consistent, users would not be able to use the app anymore.

Good point 👍 

>As for the specific exception type handling, I only know of the documented @throws in context.startService. What exceptions should we expect from FluxC queries? In only saw RuntimeException from WellSql which is not as specific as I wanted it to be. 

I'm a bit confused - this class doesn't perform any FluxC queries, right?

I think we can expect `the documented @throws in context.startService` and `CancellationException`. But I agree with your comment that we don't want to crash the app since this class will be started automatically on app-start, so we might actually want to keep the general "Exception" here.